### PR TITLE
Implement feature for functions host: support client requests

### PR DESCRIPTION
**Pull request by sebastianburckhardt**
This PR implements the feature that allows the Azure Functions Host to support client-initiated orchestrations, updates, reads, and events.

It adds a project `Counter.Service.OnFunctions` which demonstrates a counter service using HTTP bindings for "ForkIncrement" and "PerformIncrement", and modifies the HelloWorld sample to also use a client request (as opposed to running hello world on startup).

### CODE REVIEW

**Comment by cmeiklejohn** [comment]
It looks like this branch needs a rebase?  There's apparently conflicts on the HelloWorld sample.

**Comment by sebastianburckhardt** [comment]
Yep, I will need to do that in the end (but I will wait until after we are through reviewing and addressing feedback etc).

File: HelloWorld/HelloWorld.Service.OnFunctions/Functions/HelloWorld.cs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+using FunctionsHost;
+
+
+namespace HelloWorld.Service.OnFunctions
+{
+    public static class HelloWorld
+    {
+        [FunctionName("HelloWorld")]
+        public static async Task<IActionResult> Run(
+            [HttpTrigger(AuthorizationLevel.Function, "post", Route = null)] HttpRequest req,
```
Comment (cmeiklejohn) [inline HelloWorld/HelloWorld.Service.OnFunctions/Functions/HelloWorld.cs]: It looks like you made the ```Initialize``` function no longer anonymous, but shouldn't the HTTP trigger for the request be anonymous?

File: Hosts/FunctionsHost/Transport/ResponseSender.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace FunctionsHost
+{
+    internal class ResponseSender : BatchSender<IResponseMessage>
+    {
+        public const int NumberPartitions = 32;
```
Comment (cmeiklejohn) [inline Hosts/FunctionsHost/Transport/ResponseSender.cs]: Should this be a configuration parameter?

Comment (sebastianburckhardt) [reply Hosts/FunctionsHost/Transport/ResponseSender.cs]: Actually, I will leave it fixed for now. It is not safe for users to change this dynamically, and I think there is no harm in keeping the partitions at the max.

**Pull request merged**